[dynamo] implement custom dict variable as a general solution for HF's ModelOutput class

weifengpy opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105044

before the PR, for HF's ModelOutput class, we use dicts.py/DataClassVariable with our own implementation on __getItem__, __setAttr__, __setItem__. There is a risk that ModelOutput logic may change since it is a user code

after the PR, we inline __getItem__, __setAttr__, __setItem__ using dicts.py/CustomizedDictVariable so the logic always keep AA

unit test
* python test/dynamo/test_model_output.py -k test_HF_bert_model_output

test on HF benchmark
* python benchmarks/dynamo/huggingface.py -d cuda --inference --accuracy --progress --inductor --print-dataframe-summary 2>&1 
* all metric are the same before/after the PR, including pass rate, unique_graphs, graph_breaks, unique_graph_breaks
  * before the PR: P790393916
  * after the PR: P790368991 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @anijain2305 @ipiszy

weifengpy committed:
[WIP][dynamo] implement custom dict variable

[ghstack-poisoned]

weifengpy committed:
Update on "[WIP][dynamo] implement custom dict variable"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

weifengpy committed:
Update on "[WIP][dynamo] implement custom dict variable"




cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

weifengpy committed:
Update on "[dynamo] implement custom dict variable as a general solution for HF's ModelOutput class"


before the PR, for HF's ModelOutput class, we use dicts.py/DataClassVariable with our own implementation on __getItem__, __setAttr__, __setItem__. There is a risk that ModelOutput logic may change since it is a user code

after the PR, we inline __getItem__, __setAttr__, __setItem__ using dicts.py/CustomizedDictVariable so the logic always keep AA

tested the change with HF benchmark
* python benchmarks/dynamo/huggingface.py -d cuda --inference --accuracy --progress --inductor --print-dataframe-summary 2>&1 
* all metric are the same before/after the PR, including pass rate, unique_graphs, graph_breaks, unique_graph_breaks
  * before the PR: P790393916
  * after the PR: P790368991 


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

weifengpy committed:
Update on "[dynamo] implement custom dict variable as a general solution for HF's ModelOutput class"


before the PR, for HF's ModelOutput class, we use dicts.py/DataClassVariable with our own implementation on __getItem__, __setAttr__, __setItem__. There is a risk that ModelOutput logic may change since it is a user code

after the PR, we inline __getItem__, __setAttr__, __setItem__ using dicts.py/CustomizedDictVariable so the logic always keep AA

unit test
* python test/dynamo/test_model_output.py -k test_HF_bert_model_output

test on HF benchmark
* python benchmarks/dynamo/huggingface.py -d cuda --inference --accuracy --progress --inductor --print-dataframe-summary 2>&1 
* all metric are the same before/after the PR, including pass rate, unique_graphs, graph_breaks, unique_graph_breaks
  * before the PR: P790393916
  * after the PR: P790368991 


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

willfengg commented:
the CI job failure [cuda12.1-py3.10-gcc9-sm86-periodic-dynamo-benchmarks / test (dynamo_eager_torchbench, 1, 1, linux.g5.4xlarge.nvidia.gpu)](https://hud.pytorch.org/pr/pytorch/pytorch/105044#15118320075) ([gh](https://github.com/pytorch/pytorch/actions/runs/5582187203/jobs/10201915353)) looks unrelated

```
running setup for /var/lib/jenkins/workspace/torchbench/torchbenchmark/models/pytorch_struct...FAIL

Error for /var/lib/jenkins/workspace/torchbench/torchbenchmark/models/pytorch_struct:
---------------------------------------------------------------------------
  WARNING: Did not find branch or tag 'f4e374e', assuming revision or ref.
ERROR: THESE PACKAGES DO NOT MATCH THE HASHES FROM THE REQUIREMENTS FILE. If you have updated the package versions, please update the hashes. Otherwise, examine the package contents carefully; someone may have tampered with them.
    sacremoses from https://files.pythonhosted.org/packages/28/78/fef8d089db5b97546fd6d1ff2e813b8544e85670bf3a8c378c9d0250b98d/sacremoses-0.0.53.tar.gz (from pytorch-transformers->-r requirements.txt (line 3)):
        Expected sha256 43715868766c643b35de4b8046cce236bfe59a7fa88b25eaf6ddf02bacf53a7a
             Got        74de8a07efcbfcd24f183360c364e518af931ff00f40f68ced0e4ada5e5568cb
```

weifengpy committed:
Update on "[dynamo] implement custom dict variable as a general solution for HF's ModelOutput class"


before the PR, for HF's ModelOutput class, we use dicts.py/DataClassVariable with our own implementation on __getItem__, __setAttr__, __setItem__. There is a risk that ModelOutput logic may change since it is a user code

after the PR, we inline __getItem__, __setAttr__, __setItem__ using dicts.py/CustomizedDictVariable so the logic always keep AA

unit test
* python test/dynamo/test_model_output.py -k test_HF_bert_model_output

test on HF benchmark
* python benchmarks/dynamo/huggingface.py -d cuda --inference --accuracy --progress --inductor --print-dataframe-summary 2>&1 
* all metric are the same before/after the PR, including pass rate, unique_graphs, graph_breaks, unique_graph_breaks
  * before the PR: P790393916
  * after the PR: P790368991 


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

weifengpy committed:
Update on "[dynamo] implement custom dict variable as a general solution for HF's ModelOutput class"


before the PR, for HF's ModelOutput class, we use dicts.py/DataClassVariable with our own implementation on __getItem__, __setAttr__, __setItem__. There is a risk that ModelOutput logic may change since it is a user code

after the PR, we inline __getItem__, __setAttr__, __setItem__ using dicts.py/CustomizedDictVariable so the logic always keep AA

unit test
* python test/dynamo/test_model_output.py -k test_HF_bert_model_output

test on HF benchmark
* python benchmarks/dynamo/huggingface.py -d cuda --inference --accuracy --progress --inductor --print-dataframe-summary 2>&1 
* all metric are the same before/after the PR, including pass rate, unique_graphs, graph_breaks, unique_graph_breaks
  * before the PR: P790393916
  * after the PR: P790368991 


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

yanboliang commented on torch/_dynamo/variables/dicts.py:
```diff
@@ -429,6 +430,116 @@ def var_getattr(self, tx, name: str) -> "VariableTracker":
+                variables.UserFunctionVariable(fn, source=source, **options),
+                [self] + list(args),
+                kwargs,
+            )
```
I think for ```__setitem__``` and ```__setattr__```, we probably can't simple inlining them, especially if the attribute to be set is a function. cc @jansel

yanboliang commented on torch/_dynamo/variables/misc.py:
```diff
@@ -112,6 +112,29 @@ def call_method(
+            return tx.replace_all(
+                self.objvar,
+                self.objvar.modifed(newval, new_rec_contains, **options),
+            )
```
* I think this is only valid if the parent class is ```dict``` or ```OrderedDict```, what about if the parent class is another customized dict class.
* I guess this may trigger side effect.

willfengg replied:
> parent class is dict or OrderedDict

I updated the diff to check ```inner_fn in (collections.OrderedDict.__setitem__, object.__setattr__)```

weifengpy committed:
Update on "[dynamo] implement custom dict variable as a general solution for HF's ModelOutput class"


before the PR, for HF's ModelOutput class, we use dicts.py/DataClassVariable with our own implementation on __getItem__, __setAttr__, __setItem__. There is a risk that ModelOutput logic may change since it is a user code

after the PR, we inline __getItem__, __setAttr__, __setItem__ using dicts.py/CustomizedDictVariable so the logic always keep AA

unit test
* python test/dynamo/test_model_output.py -k test_HF_bert_model_output

test on HF benchmark
* python benchmarks/dynamo/huggingface.py -d cuda --inference --accuracy --progress --inductor --print-dataframe-summary 2>&1 
* all metric are the same before/after the PR, including pass rate, unique_graphs, graph_breaks, unique_graph_breaks
  * before the PR: P790393916
  * after the PR: P790368991 


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

jansel commented on torch/_dynamo/variables/builtin.py:
```diff
@@ -1104,6 +1104,8 @@ def call_setattr(
     ):
         if isinstance(obj, variables.DataClassVariable):
             return obj.call_method(tx, "__setattr__", [name_var, val], {})
+        if isinstance(obj, variables.CustomizedDictVariable):
+            return obj.call_method(tx, "__setattr__", [name_var, val], {})
```
```suggestion
        if isinstance(obj, (variables.DataClassVariable, variables.CustomizedDictVariable)):
            return obj.call_method(tx, "__setattr__", [name_var, val], {})
```

willfengg replied:
merged ifs

jansel commented on torch/_dynamo/variables/dicts.py:
```diff
@@ -429,6 +430,116 @@ def var_getattr(self, tx, name: str) -> "VariableTracker":
 
+class CustomizedDictVariable(ConstDictVariable):
+    @staticmethod
+    @functools.lru_cache(None)
```
Using `@functools.lru_cache(None)` here might cause leaks since it will hold a pointer to the user_cls arg.

User classes can get created inside a local scope, so it is possible to create them in a loop.

willfengg replied:
thanks. I removed lru_cache for now. I can improve if there are any better solutions

jansel commented on torch/_dynamo/variables/dicts.py:
```diff
@@ -429,6 +430,116 @@ def var_getattr(self, tx, name: str) -> "VariableTracker":
+    # called from user_defined.py
+    # when is_matching_cls(cls) is true
+    @classmethod
+    def create(cls, user_cls, args, kwargs, options):
```
This function seems to be making assumptions about the behavior of `user_cls.__init__` (and maybe post_init).  If `__init__` is the default implementation, this is fine, but if it is user code it likely isn't.  We should have some checks on the behavior of `__init__`.

willfengg replied:
I added __init__ and __post_init__ check in ```is_matching_cls```
* return True if OrderedDict.__init__
* return True if ssubclass(cls, ModelOutput). __init__ and __post_init__ are overriden. here we still have assumptions on no big changes in HF ModelOutput initialization as it is user code

I am open to change if there are better ways to make it more generalizable

weifengpy committed:
Update on "[dynamo] implement custom dict variable as a general solution for HF's ModelOutput class"


before the PR, for HF's ModelOutput class, we use dicts.py/DataClassVariable with our own implementation on __getItem__, __setAttr__, __setItem__. There is a risk that ModelOutput logic may change since it is a user code

after the PR, we inline __getItem__, __setAttr__, __setItem__ using dicts.py/CustomizedDictVariable so the logic always keep AA

unit test
* python test/dynamo/test_model_output.py -k test_HF_bert_model_output

test on HF benchmark
* python benchmarks/dynamo/huggingface.py -d cuda --inference --accuracy --progress --inductor --print-dataframe-summary 2>&1 
* all metric are the same before/after the PR, including pass rate, unique_graphs, graph_breaks, unique_graph_breaks
  * before the PR: P790393916
  * after the PR: P790368991 


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

weifengpy committed:
Update on "[dynamo] implement custom dict variable as a general solution for HF's ModelOutput class"


before the PR, for HF's ModelOutput class, we use dicts.py/DataClassVariable with our own implementation on __getItem__, __setAttr__, __setItem__. There is a risk that ModelOutput logic may change since it is a user code

after the PR, we inline __getItem__, __setAttr__, __setItem__ using dicts.py/CustomizedDictVariable so the logic always keep AA

unit test
* python test/dynamo/test_model_output.py -k test_HF_bert_model_output

test on HF benchmark
* python benchmarks/dynamo/huggingface.py -d cuda --inference --accuracy --progress --inductor --print-dataframe-summary 2>&1 
* all metric are the same before/after the PR, including pass rate, unique_graphs, graph_breaks, unique_graph_breaks
  * before the PR: P790393916
  * after the PR: P790368991 


cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78

[ghstack-poisoned]

weifengpy commented:
@pytorchmergebot merge
